### PR TITLE
global: setup automatic updates and releases

### LINKED ISSUE #7: global: setup automatic updates and releases
Since this repository uses the lists from https://github.com/atmire/COUNTER-Robots/, syncing should be done in an automatic fashion (e.g. on Travis) and be released in regular intervals (e.g. each month).

### PR DESCRIPTION
**Pull request by jorikvankemenade**
This PR introduces two major changes: the update script and the Travis builds.

## Update script
We use the lists from [Making Data Count](https://github.com/CDLUC3/Make-Data-Count/tree/master/user-agents) for distinguishing user agents that are either robots or machines. These lists are a manually curated division of the lists from [Counter Robots](https://github.com/atmire/COUNTER-Robots/). The problem is that the original list is updated more often than the manually curated one. To mitigate this we updated the update script to always use the latest “Counter Robots” version, and use the “Making Data Count” lists for curating the `robots.txt`.

## Travis builds
Since we can automatically sync with the external libraries it is also nice to automatically release an updated version of the user-agents list. This is done by creating some additional build stages in Travis that can be used for automatically releasing updated versions.

This PR distinguishes three kinds of builds:
1. Regular build
2. Cron triggered build
3. Tagged build

**Regular build**
For the regular build nothing changes. It runs the tests for Python versions 2.7, 3.5 and 3.6.

image

**Cron triggered build**
We can set a `cron` build in Travis that triggers every month. This build will run the `update_script`. If changes are detected the build is tested and if this is successful the changes are committed, tagged and pushed. This will automatically cause a Tagged build.

image

**Tagged build**
A tagged build is like a regular build, but with an additional stage. If the `test` stage is successful a deploy stage will run that will create a deploy in pypi. This allows us to release builds manually, by tagging a release, or via the cron triggered builds.

image

## Before merging
Before we merge this PR we must update the secret in `.travis.yml`. This should be set to the Github token that we have for Invenio i.e. `GH_TOKEN=<token>`. 

## After merging
After we have merged this PR we should enable a monthly cron build on the master branch. This will check monthly if there are any updates. If there are updates we will release a new version.

Closes: #7

### CODE REVIEW

File: .editorconfig
Comment (slint) [inline .editorconfig]: We usually extend the copyright years to include the entire lifetime of the project, i.e. `2018-2019`

File: run-update.sh
Comment (slint) [inline run-update.sh]: These can be passed as a parameter, like `git commit --author='Invenio <info@inveniosoftware.org>' ...`

File: run-update.sh
Comment (slint) [inline run-update.sh]: The `v` is not needed for prefixing the version: https://github.com/inveniosoftware/counter-robots/blob/484943fdc7e08f41d3ad7a9e2229afe0cec05547/counter_robots/version.py#L17

File: run-update.sh
Comment (slint) [inline run-update.sh]:
```suggestion
    echo "Tag $TAG already exists"
```

File: run-update.sh
Comment (slint) [inline run-update.sh]:
```suggestion
    echo "Changes found. Commiting and releasing v$TAG."
```

File: run-update.sh
Comment (slint) [inline run-update.sh]: In general, the commit structure/workflow should be the same as described in https://github.com/inveniosoftware/counter-robots/blob/master/docs/developer.rst

- One commit for the change
- One commit for the release (which is also tagged)

File: .travis.yml
Comment (slint) [inline .travis.yml:51]: Ususally we keep the `deploy` stage [on the top level](https://github.com/inveniosoftware/invenio-records-rest/blob/16c7af653ff720e78d30b64d26ec15951ce4eaca/.travis.yml#L73-L82). I guess it works like this, but for the sake of consistency when the next person comes across the module

**Review by slint — COMMENTED** [review]
LGTM :+1: Some minor changes and then we have to coordinate adding tokens etc.

File: .travis.yml
Comment (jorikvankemenade) [reply .travis.yml:51]: It depends a bit on what behaviour we want. If we use this we run a separate deploy _stage_ that includes a deploy _step_ for every tagged build, i.e.:

image

All other builds don't trigger the deploy _step_. The advantage of this approach is that running a deploy is an explicit thing, not something that is buried in the logs of every job.

If we remove the deploy _stage_ than we will have  the same behaviour we had before, i.e. every  Travis job will run the deploy step and check if it might be a release or not.

File: run-update.sh
Comment (jorikvankemenade) [reply run-update.sh]: Okay, the last version is `v2018.6` so I assumed that we add the v. 

```
Jorik:counter-robots/ (feature/automatic-release✗) $ git tag                              [14:44:58]
v2018.6
v2018.6a1
```

Comment (slint) [reply run-update.sh]: Indeed we do... But it's only for the tag (because of historical reasons). The `__version__` should be unprefixed

File: run-update.sh
Comment (jorikvankemenade) [reply run-update.sh]: Unfortunately, it is not possible to create an annotated tag using the `--author` tag. So for a correct signing of the tag we need to explicitly set the name and email of the Git user.

**Comment by jorikvankemenade** [comment]
@slint I have applied your suggested changes. I am not sure about your comment about the deploy, but maybe we can quickly discuss that offline. 

You also linked `docs/developers.rst` these might be obsolete. Do we want to remove them, or describe the new procedure there?

File: run-update.sh
Comment (jorikvankemenade) [reply run-update.sh]: So today's version would be `2019.07.25` but the tag would be `v2019.07.25`?

**Comment by jorikvankemenade** [comment]
@slint I squashed my changes. Please let me know if I messed up the format or anything. A gentle reminder that we need to change the secret in `.travis.yml` before merging.

File: push-update.sh
Comment (slint) [inline push-update.sh]: Oof, missed that one. I think Travis has a bunch of environment variables for e.g. the PRs GitHub repo: https://docs.travis-ci.com/user/environment-variables/#default-environment-variables

File: run-update.sh
Comment (slint) [inline run-update.sh]:
```suggestion
    git commit -m "release: $VERSION"
```

**Review by slint — APPROVED** [review]

**Comment by jorikvankemenade** [comment]
@slint @lnielsen this is ready for merging as soon as we changed the `GH_TOKEN` in `.travis.yml`.

File: run-update.sh
Comment (jirikuncar) [inline run-update.sh]:
```suggestion
# Commit all changes for the TRAVIS_BUILD_NUMBER
```

Comment (jorikvankemenade) [reply run-update.sh]: Thanks for pointing this out!

**Review by topless — APPROVED** [review]

**Comment by slint** [comment]
I'm changing the "update" Travis stage to do the `git push ...` via using GitHub deploy keys instead of the encrypted GitHub token. 

The reason is that the GH token would have to belong to one of the `inveniosoftware` admins, and thus would have permissions to modify any repository (there's no way to limit the scope of a GH personal access token). Deploy keys provide write access on the repository-level, which is preferred (and doesn't also depend on an actual `inveniosoftware` org member).

**Pull request closed without merging**
